Improvement: MIDI emulator backend (sine synth) for platforms that don't support MIDI otherwise

dgruss opened the pull request:
This PR makes it possible to build and run USDX with synthesized, emulated MIDI output instead of using an actual MIDI audio device. The MIDI emulation backend is compile-time-selectable (`--enable-midiemu`) for systems without native MIDI (e.g., Raspberry Pi, macOS??). It synthesizes a single-voice sine wave with simple ADSR envelope, exposed behind the existing `MidiOut` interface. No changes required for existing code using `TMidiOutput`. It also fixes note-off logic in the editor: notes now stop exactly at their end, even with gaps, and repeated notes retrigger cleanly.

All changes are minimal and isolated; existing MIDI backends are untouched and the new code is only activated with `--enable-midiemu`.

dgruss committed:
add a midi emulation for devices without midi support

s09bQ5 reviewed (commented):
Calling FPlaybackStream.Play for every note causes clicking because it flushes the stream buffers. To get really clean note starts it is not necessary to reset the phase. It is more important to make your FRelease code work again.

After disabling the Stop() and repeated Play() as well as the FPhase resetting, I have experimented with adding an exponentially decaying overtone to make it possible to hear when a note is restarted at the same pitch. It sounds a bit like a banjo. Hammond organs do the same for their percussive tone feature (and they also don't reset the phase when starting a tone).

I think we do need a mutex to prevent that HandleMidiEvent changes the parameters while ReadBuffer runs.

s09bQ5 commented on src/lib/midiemu/MidiAudioSourceStream.pas:
```diff
@@ -0,0 +1,187 @@
+  begin
+    FNote := Data1;
+    FFreq := 440.0 * Power(2.0, (FNote - 69)/12.0);
+    FVel := Data2 / 127.0;
```
[This paper](https://www.cs.cmu.edu/~rbd/papers/velocity-icmc2006.pdf) says that many real midi instruments actually use a different formula. Their amplitude is roughly proportional to the square of the velocity. It also proposes to use 1/1000th of the maximum amplitude for the minimal velocity.

s09bQ5 commented on src/lib/midiemu/MidiAudioSourceStream.pas:
```diff
@@ -0,0 +1,238 @@
+  phaseInc := 2 * Pi * FFreq / SampleRate;
+  overtoneInc := 4 * Pi * FFreq / SampleRate; // 2nd harmonic
+
+  FLock.Enter;
```
FLock.Enter also needs to protect FFreq.

s09bQ5 commented on src/lib/midiemu/MidiAudioSourceStream.pas:
```diff
@@ -0,0 +1,238 @@
+        // FPhase handling: keep between -Pi and +Pi
+        FPhase := FPhase + phaseInc;
+        if FPhase > Pi then FPhase := FPhase - 2 * Pi;
+        if FPhase < -Pi then FPhase := FPhase + 2 * Pi;
```
FPhase will never be < -Pi because phaseInc is always positive and when we subtract from FPhase in line 151, it ends up being > -Pi.

dgruss commented:
Traveling right now, I gave you access to that branch

s09bQ5 committed:
midiemu: no need to track overtone phase

It is twice the phase of the fundamental.

s09bQ5 committed:
midiemu: protect access to FFreq with mutex

s09bQ5 committed:
midiemu: FPhase can't be smaller than -Pi

s09bQ5 committed:
SoftMixer: do nothing when Play is called on playing stream

s09bQ5 committed:
midiemu: avoid FEnvLevel ramping up while FActive false

FActive will still be set to false inside the loop, but FRelease will
remain true until the next note is played.

s09bQ5 committed:
midiemu: don't force FEnvLevel to zero on new note

s09bQ5 committed:
midiemu: ignore note off for wrong note

s09bQ5 committed:
midiemu: make overtone decay independent of sample rate

s09bQ5 committed:
midiemu: implement volume control

To make the slider in the upper right of the editor work.

s09bQ5 commented:
Made some changes, please check.

I'm aware that the timing is far from precise, but for a first version this should be enough.

dgruss commented:
That affects the note end right? Not the note start?

dgruss commented:
I was thinking whether we could empty/flush the buffer or something when a new note is to be started AND the stream is silent

dgruss commented:
Or if we could generate things such they have a constant offset, eg 50ms then we could fix that with #1109

s09bQ5 commented:
> I was thinking whether we could empty/flush the buffer or something when a new note is to be started AND the stream is silent

Which buffer? The SOURCE_BUFFER_FRAMES sized one? TGenericPlaybackStream.ReadData is most likely called with a bigger buffer, so that the SOURCE_BUFFER_FRAMES sized one ends up empty.

> Or if we could generate things such they have a constant offset, eg 50ms then we could fix that with https://github.com/UltraStar-Deluxe/USDX/pull/1109

The offset will always be the same on a machine as long as the sound settings are the same. It should be equal to the size of the whole buffer used by the sound driver because TAudioMixerStream.ReadData is called when the sound driver is done playing a chunk (in ALSA speak "period") of the buffer and wants to refill it.

dgruss commented:
Not sure about the constant delay right now. Measuring from putShort to readData already has a lot of variance if I see it correctly

s09bQ5 commented:
The constant delay is from ReadData being called to the samples leaving the speaker. If we switch to the approach outlined above, the delay from putShort to the first sample of the note leaving the speaker would be constant as well.

dgruss commented:
this PR also sits around for a while now with approval from one of our mac users ;)

dgruss commented:
@s09bQ5 i would make a nice commit message, force-push so that it can be merged, is that okay with you?

barbeque-squared reviewed (approved):
this works fine on my end. can't comment on actual midi devices or windows, but I'll approve and merge anyway because I'm not seeing anything that should actively cause issues in those situations either
